### ISSUE TITLE
Crash in 1.1.54265 when dropping tables

### ISSUE THREAD

**Issue by vavrusa** [yandex/ClickHouse#1128]
One instance crashed when dropping a bunch of tables, I suspect it's because I dropped the ".inner." table before the drop of MV was finished.

```
2017-08-21T03:20:01.385 36s1000 pure virtual method called
[ 331 ] <Error> BaseDaemon: (from thread 330) Terminate called without an active exception
2017-08-21T03:20:01.386 36s1000 
[ 331 ] <Error> BaseDaemon: ########################################
[ 331 ] <Error> BaseDaemon: (from thread 330) Received signal Aborted (6).
[ 331 ] <Error> BaseDaemon: 0. /lib/x86_64-linux-gnu/libc.so.6(gsignal+0xcf) [0x7fb2bf01bfcf]
[ 331 ] <Error> BaseDaemon: 1. /lib/x86_64-linux-gnu/libc.so.6(abort+0x16a) [0x7fb2bf01d3fa]
[ 331 ] <Error> BaseDaemon: 2. /usr/bin/clickhouse-server() [0x118fb4d]
[ 331 ] <Error> BaseDaemon: 3. /usr/bin/clickhouse-server(__cxxabiv1::__terminate(void (*)())+0x6) [0x36fd856]
[ 331 ] <Error> BaseDaemon: 4. /usr/bin/clickhouse-server() [0x36fd8a1]
[ 331 ] <Error> BaseDaemon: 5. /usr/bin/clickhouse-server() [0x36fd13f]
[ 331 ] <Error> BaseDaemon: 6. /usr/bin/clickhouse-server(DB::ITableDeclaration::getColumnsList() const+0x29) [0x2b3d859]
[ 331 ] <Error> BaseDaemon: 7. /usr/bin/clickhouse-server(DB::ITableDeclaration::check(DB::Block const&, bool) const+0x42) [0x2b46e42]
[ 331 ] <Error> BaseDaemon: 8. /usr/bin/clickhouse-server(DB::MergeTreeDataWriter::splitBlockIntoParts[abi:cxx11](DB::Block const&)+0x44) [0x2c33494]
[ 331 ] <Error> BaseDaemon: 9. /usr/bin/clickhouse-server(DB::ReplicatedMergeTreeBlockOutputStream::write(DB::Block const&)+0x98) [0x2c66158]
[ 331 ] <Error> BaseDaemon: 10. /usr/bin/clickhouse-server(DB::PushingToViewsBlockOutputStream::write(DB::Block const&)+0x45d) [0x29dbc5d]
[ 331 ] <Error> BaseDaemon: 11. /usr/bin/clickhouse-server(DB::copyData(DB::IBlockInputStream&, DB::IBlockOutputStream&, std::atomic<bool>*)+0xb5) [0x2df8355]
[ 331 ] <Error> BaseDaemon: 12. /usr/bin/clickhouse-server(DB::PushingToViewsBlockOutputStream::write(DB::Block const&)+0x409) [0x29dbc09]
[ 331 ] <Error> BaseDaemon: 13. /usr/bin/clickhouse-server(DB::MaterializingBlockOutputStream::write(DB::Block const&)+0x37) [0x29d9127]
[ 331 ] <Error> BaseDaemon: 14. /usr/bin/clickhouse-server(DB::AddingDefaultBlockOutputStream::write(DB::Block const&)+0x281) [0x2d259b1]
[ 331 ] <Error> BaseDaemon: 15. /usr/bin/clickhouse-server(DB::ProhibitColumnsBlockOutputStream::write(DB::Block const&)+0x5c) [0x2dc59ac]
[ 331 ] <Error> BaseDaemon: 16. /usr/bin/clickhouse-server(DB::SquashingBlockOutputStream::write(DB::Block const&)+0x3c9) [0x2dcd169]
[ 331 ] <Error> BaseDaemon: 17. /usr/bin/clickhouse-server(DB::CountingBlockOutputStream::write(DB::Block const&)+0x2b) [0x2d4bc0b]
[ 331 ] <Error> BaseDaemon: 18. /usr/bin/clickhouse-server(DB::TCPHandler::receiveData()+0xae) [0x10b792e]
[ 331 ] <Error> BaseDaemon: 19. /usr/bin/clickhouse-server(DB::TCPHandler::receivePacket()+0x1e1) [0x10b8041]
[ 331 ] <Error> BaseDaemon: 20. /usr/bin/clickhouse-server(DB::TCPHandler::readData(DB::Settings const&)+0x1a9) [0x10b83c9]
[ 331 ] <Error> BaseDaemon: 21. /usr/bin/clickhouse-server(DB::TCPHandler::processInsertQuery(DB::Settings const&)+0x274) [0x10b86f4]
[ 331 ] <Error> BaseDaemon: 22. /usr/bin/clickhouse-server(DB::TCPHandler::runImpl()+0x6b9) [0x10b8f39]
[ 331 ] <Error> BaseDaemon: 23. /usr/bin/clickhouse-server(DB::TCPHandler::run()+0x2b) [0x10b9aab]
[ 331 ] <Error> BaseDaemon: 24. /usr/bin/clickhouse-server(Poco::Net::TCPServerConnection::start()+0xf) [0x3329bbf]
[ 331 ] <Error> BaseDaemon: 25. /usr/bin/clickhouse-server(Poco::Net::TCPServerDispatcher::run()+0x13b) [0x332b40b]
[ 331 ] <Error> BaseDaemon: 26. /usr/bin/clickhouse-server(Poco::PooledThread::run()+0xb7) [0x3448fc7]
[ 331 ] <Error> BaseDaemon: 27. /usr/bin/clickhouse-server(Poco::ThreadImpl::runnableEntry(void*)+0xa5) [0x34432b5]
[ 331 ] <Error> BaseDaemon: 28. /lib/x86_64-linux-gnu/libpthread.so.0(+0x7494) [0x7fb2bfa9f494]
2017-08-21T03:20:01.389 36s1000 pure virtual method called
[ 331 ] <Error> BaseDaemon: (from thread 323) Terminate called without an active exception
2017-08-21T03:20:01.389 36s1000 
[ 331 ] <Error> BaseDaemon: ########################################
[ 331 ] <Error> BaseDaemon: (from thread 323) Received signal Aborted (6).
[ 331 ] <Error> BaseDaemon: 0. /lib/x86_64-linux-gnu/libc.so.6(gsignal+0xcf) [0x7fb2bf01bfcf]
[ 331 ] <Error> BaseDaemon: 1. /lib/x86_64-linux-gnu/libc.so.6(abort+0x16a) [0x7fb2bf01d3fa]
[ 331 ] <Error> BaseDaemon: 2. /usr/bin/clickhouse-server() [0x118fb4d]
[ 331 ] <Error> BaseDaemon: 3. /usr/bin/clickhouse-server(__cxxabiv1::__terminate(void (*)())+0x6) [0x36fd856]
[ 331 ] <Error> BaseDaemon: 4. /usr/bin/clickhouse-server() [0x36fd8a1]
[ 331 ] <Error> BaseDaemon: 5. /usr/bin/clickhouse-server() [0x36fd13f]
[ 331 ] <Error> BaseDaemon: 6. /usr/bin/clickhouse-server(DB::ITableDeclaration::getColumnsList() const+0x29) [0x2b3d859]
[ 331 ] <Error> BaseDaemon: 7. /usr/bin/clickhouse-server(DB::ITableDeclaration::check(DB::Block const&, bool) const+0x42) [0x2b46e42]
[ 331 ] <Error> BaseDaemon: 8. /usr/bin/clickhouse-server(DB::MergeTreeDataWriter::splitBlockIntoParts[abi:cxx11](DB::Block const&)+0x44) [0x2c33494]
[ 331 ] <Error> BaseDaemon: 9. /usr/bin/clickhouse-server(DB::ReplicatedMergeTreeBlockOutputStream::write(DB::Block const&)+0x98) [0x2c66158]
[ 331 ] <Error> BaseDaemon: 10. /usr/bin/clickhouse-server(DB::PushingToViewsBlockOutputStream::write(DB::Block const&)+0x45d) [0x29dbc5d]
[ 331 ] <Error> BaseDaemon: 11. /usr/bin/clickhouse-server(DB::copyData(DB::IBlockInputStream&, DB::IBlockOutputStream&, std::atomic<bool>*)+0xb5) [0x2df8355]
[ 331 ] <Error> BaseDaemon: 12. /usr/bin/clickhouse-server(DB::PushingToViewsBlockOutputStream::write(DB::Block const&)+0x409) [0x29dbc09]
[ 331 ] <Error> BaseDaemon: 13. /usr/bin/clickhouse-server(DB::MaterializingBlockOutputStream::write(DB::Block const&)+0x37) [0x29d9127]
[ 331 ] <Error> BaseDaemon: 14. /usr/bin/clickhouse-server(DB::AddingDefaultBlockOutputStream::write(DB::Block const&)+0x281) [0x2d259b1]
[ 331 ] <Error> BaseDaemon: 15. /usr/bin/clickhouse-server(DB::ProhibitColumnsBlockOutputStream::write(DB::Block const&)+0x5c) [0x2dc59ac]
[ 331 ] <Error> BaseDaemon: 16. /usr/bin/clickhouse-server(DB::SquashingBlockOutputStream::write(DB::Block const&)+0x3c9) [0x2dcd169]
[ 331 ] <Error> BaseDaemon: 17. /usr/bin/clickhouse-server(DB::CountingBlockOutputStream::write(DB::Block const&)+0x2b) [0x2d4bc0b]
[ 331 ] <Error> BaseDaemon: 18. /usr/bin/clickhouse-server(DB::TCPHandler::receiveData()+0xae) [0x10b792e]
[ 331 ] <Error> BaseDaemon: 19. /usr/bin/clickhouse-server(DB::TCPHandler::receivePacket()+0x1e1) [0x10b8041]
[ 331 ] <Error> BaseDaemon: 20. /usr/bin/clickhouse-server(DB::TCPHandler::readData(DB::Settings const&)+0x1a9) [0x10b83c9]
[ 331 ] <Error> BaseDaemon: 21. /usr/bin/clickhouse-server(DB::TCPHandler::processInsertQuery(DB::Settings const&)+0x274) [0x10b86f4]
[ 331 ] <Error> BaseDaemon: 22. /usr/bin/clickhouse-server(DB::TCPHandler::runImpl()+0x6b9) [0x10b8f39]
[ 331 ] <Error> BaseDaemon: 23. /usr/bin/clickhouse-server(DB::TCPHandler::run()+0x2b) [0x10b9aab]
[ 331 ] <Error> BaseDaemon: 24. /usr/bin/clickhouse-server(Poco::Net::TCPServerConnection::start()+0xf) [0x3329bbf]
[ 331 ] <Error> BaseDaemon: 25. /usr/bin/clickhouse-server(Poco::Net::TCPServerDispatcher::run()+0x13b) [0x332b40b]
[ 331 ] <Error> BaseDaemon: 26. /usr/bin/clickhouse-server(Poco::PooledThread::run()+0xb7) [0x3448fc7]
[ 331 ] <Error> BaseDaemon: 27. /usr/bin/clickhouse-server(Poco::ThreadImpl::runnableEntry(void*)+0xa5) [0x34432b5]
[ 331 ] <Error> BaseDaemon: 28. /lib/x86_64-linux-gnu/libpthread.so.0(+0x7494) [0x7fb2bfa9f494]
2017-08-21T03:20:01.391 36s1000 pure virtual method called
[ 331 ] <Error> BaseDaemon: (from thread 332) Terminate called without an active exception
2017-08-21T03:20:01.391 36s1000 
[ 331 ] <Error> BaseDaemon: ########################################
[ 331 ] <Error> BaseDaemon: (from thread 332) Received signal Aborted (6).
[ 331 ] <Error> BaseDaemon: 0. /lib/x86_64-linux-gnu/libc.so.6(gsignal+0xcf) [0x7fb2bf01bfcf]
[ 331 ] <Error> BaseDaemon: 1. /lib/x86_64-linux-gnu/libc.so.6(abort+0x16a) [0x7fb2bf01d3fa]
[ 331 ] <Error> BaseDaemon: 2. /usr/bin/clickhouse-server() [0x118fb4d]
[ 331 ] <Error> BaseDaemon: 3. /usr/bin/clickhouse-server(__cxxabiv1::__terminate(void (*)())+0x6) [0x36fd856]
[ 331 ] <Error> BaseDaemon: 4. /usr/bin/clickhouse-server() [0x36fd8a1]
[ 331 ] <Error> BaseDaemon: 5. /usr/bin/clickhouse-server() [0x36fd13f]
[ 331 ] <Error> BaseDaemon: 6. /usr/bin/clickhouse-server(DB::ITableDeclaration::getColumnsList() const+0x29) [0x2b3d859]
[ 331 ] <Error> BaseDaemon: 7. /usr/bin/clickhouse-server(DB::ITableDeclaration::check(DB::Block const&, bool) const+0x42) [0x2b46e42]
[ 331 ] <Error> BaseDaemon: 8. /usr/bin/clickhouse-server(DB::MergeTreeDataWriter::splitBlockIntoParts[abi:cxx11](DB::Block const&)+0x44) [0x2c33494]
[ 331 ] <Error> BaseDaemon: 9. /usr/bin/clickhouse-server(DB::ReplicatedMergeTreeBlockOutputStream::write(DB::Block const&)+0x98) [0x2c66158]
[ 331 ] <Error> BaseDaemon: 10. /usr/bin/clickhouse-server(DB::PushingToViewsBlockOutputStream::write(DB::Block const&)+0x45d) [0x29dbc5d]
[ 331 ] <Error> BaseDaemon: 11. /usr/bin/clickhouse-server(DB::copyData(DB::IBlockInputStream&, DB::IBlockOutputStream&, std::atomic<bool>*)+0xb5) [0x2df8355]
[ 331 ] <Error> BaseDaemon: 12. /usr/bin/clickhouse-server(DB::PushingToViewsBlockOutputStream::write(DB::Block const&)+0x409) [0x29dbc09]
[ 331 ] <Error> BaseDaemon: 13. /usr/bin/clickhouse-server(DB::MaterializingBlockOutputStream::write(DB::Block const&)+0x37) [0x29d9127]
[ 331 ] <Error> BaseDaemon: 14. /usr/bin/clickhouse-server(DB::AddingDefaultBlockOutputStream::write(DB::Block const&)+0x281) [0x2d259b1]
[ 331 ] <Error> BaseDaemon: 15. /usr/bin/clickhouse-server(DB::ProhibitColumnsBlockOutputStream::write(DB::Block const&)+0x5c) [0x2dc59ac]
[ 331 ] <Error> BaseDaemon: 16. /usr/bin/clickhouse-server(DB::SquashingBlockOutputStream::write(DB::Block const&)+0x3c9) [0x2dcd169]
[ 331 ] <Error> BaseDaemon: 17. /usr/bin/clickhouse-server(DB::CountingBlockOutputStream::write(DB::Block const&)+0x2b) [0x2d4bc0b]
[ 331 ] <Error> BaseDaemon: 18. /usr/bin/clickhouse-server(DB::TCPHandler::receiveData()+0xae) [0x10b792e]
[ 331 ] <Error> BaseDaemon: 19. /usr/bin/clickhouse-server(DB::TCPHandler::receivePacket()+0x1e1) [0x10b8041]
[ 331 ] <Error> BaseDaemon: 20. /usr/bin/clickhouse-server(DB::TCPHandler::readData(DB::Settings const&)+0x1a9) [0x10b83c9]
[ 331 ] <Error> BaseDaemon: 21. /usr/bin/clickhouse-server(DB::TCPHandler::processInsertQuery(DB::Settings const&)+0x274) [0x10b86f4]
[ 331 ] <Error> BaseDaemon: 22. /usr/bin/clickhouse-server(DB::TCPHandler::runImpl()+0x6b9) [0x10b8f39]
[ 331 ] <Error> BaseDaemon: 23. /usr/bin/clickhouse-server(DB::TCPHandler::run()+0x2b) [0x10b9aab]
[ 331 ] <Error> BaseDaemon: 24. /usr/bin/clickhouse-server(Poco::Net::TCPServerConnection::start()+0xf) [0x3329bbf]
[ 331 ] <Error> BaseDaemon: 25. /usr/bin/clickhouse-server(Poco::Net::TCPServerDispatcher::run()+0x13b) [0x332b40b]
[ 331 ] <Error> BaseDaemon: 26. /usr/bin/clickhouse-server(Poco::PooledThread::run()+0xb7) [0x3448fc7]
[ 331 ] <Error> BaseDaemon: 27. /usr/bin/clickhouse-server(Poco::ThreadImpl::runnableEntry(void*)+0xa5) [0x34432b5]
[ 331 ] <Error> BaseDaemon: 28. /lib/x86_64-linux-gnu/libpthread.so.0(+0x7494) [0x7fb2bfa9f494]
```

cc @bobrik

**Comment by ztlpn** [comment]
Thanks, I was able to reproduce - the problem is that the INSERT query does not acquire lock for the .inner. table, so it could be dropped while the insertion was in progress.

**Comment by jackpgao** [comment]
Does this bug have been fixed?

**Comment by alexey-milovidov** [comment]
No. @CurtizJ is working now on this issue.

**Comment by CurtizJ** [comment]
Fixed in master #3404.
